component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/mobileconnect/MobileConnectAuthenticator.java

```diff
@@ -1082,27 +1085,55 @@ private HttpURLConnection msisdnBasedDiscoveryProcess(String authorizationHeader
+            String claimUri, claimValue;
+
             for (Map.Entry<String, Object> entry : userClaims.entrySet()) {
-                claims.put(ClaimMapping.build(entry.getKey(), entry.getKey(), null,
-                        false), entry.getValue().toString());
+                claimUri = getEffectiveClaimUri(claimDialectUri, entry.getKey(),
+                        shouldPrefixClaimDialectUri);
+                claimValue = entry.getValue().toString();
+
+                claims.put(ClaimMapping.build(claimUri, claimUri, null,
+                        false), claimValue);
+
                 if (log.isDebugEnabled() &&
                         IdentityUtil.isTokenLoggable(IdentityConstants.IdentityTokens.USER_CLAIMS)) {
-                    log.debug("Adding claim mapping : " + entry.getKey() + " <> " + entry.getKey() + " : "
-                            + entry.getValue());
+                    log.debug("Adding claim mapping : " + claimUri + " <> " + claimUri + " : "
+                            + claimValue);
                 }
             }
-            if (StringUtils.isBlank(context.getExternalIdP().getIdentityProvider().getClaimConfig().
-                    getUserClaimURI())) {
+
+            if (isUserClaimUriBlank) {
+                //Set id as the default userClaimUri in IDP claim Configs.
                 context.getExternalIdP().getIdentityProvider().getClaimConfig().setUserClaimURI
-                        (MCAuthenticatorConstants.CLAIM_ID);
+                        (getEffectiveClaimUri(claimDialectUri,
+                                MCAuthenticatorConstants.DEFAULT_USER_IDENTIFIER, shouldPrefixClaimDialectUri));
+            }
+
+            String subjectFromClaims = null;
+            if (!isUserClaimUriBlank && StringUtils.isNotBlank(claimDialectUri)) {
+                setSubject(context, userClaims);
+                context.getSubject().setUserAttributes(claims);
+                try {
+                    subjectFromClaims = FrameworkUtils.getFederatedSubjectFromClaims(context, claimDialectUri);
+                } catch (FrameworkException ex) {
+                    log.error("Error while retrieving subject from claims. " +
+                            "Both Dedicated claim dialect (" + claimDialectUri +
+                            ") and user ID Claim URI (" + userClaimUri + ") is Configured.", ex);
+                }
+            } else {
+                subjectFromClaims = FrameworkUtils.getFederatedSubjectFromClaims(
+                        context.getExternalIdP().getIdentityProvider(), claims);
             }
-            String subjectFromClaims = FrameworkUtils.getFederatedSubjectFromClaims(
-                    context.getExternalIdP().getIdentityProvider(), claims);
-            if (subjectFromClaims != null && !subjectFromClaims.isEmpty()) {
+
+            if (StringUtils.isNotBlank(subjectFromClaims)) {
                 AuthenticatedUser authenticatedUser =
```

Comment:
Did you check whether you are getting claims after you execute this line ?

Reply:
Assuming you are referring to subjectClaim.
It gives the correct subjectClaim value according to the IDP configs.

Reply:
I am not talking about subject claims. I am talking about user attributes. User attributes and subject claim are two things.

Reply:
I don't see this line do anything related to claims.
Anyways userAttributes are set at [here.](https://github.com/wso2-extensions/identity-outbound-auth-oidc-mobileconnect/pull/12/files#diff-2b4fb4b8d00ab008eeb951d6272a65e0R1143) 
Afterwards, user attributes are there in the context against the Authenticated user.

Reply:
You sent authenticated user twice isn't it ? and also user attributes twice (refer [here](https://github.com/wso2-extensions/identity-outbound-auth-oidc-mobileconnect/pull/12/files#diff-2b4fb4b8d00ab008eeb951d6272a65e0R1123))

Instead can't u just change the subject identifier of the previously created use ?

Reply:
This was once raised by Supun. Following is the reason why I didn't go for it.
"at 1143 setting of subject UserAttributes is centralized for all the cases.
removing that and distributing it introduces unnecessary complexity to the code.
Since setAttribute just set the reference there is no perf gain as well.
In conclusion, better keep the implementation this way."

Reply:
Then better to add a comment on the above line explaining why you want to add claims twice. Otherwise someone may think this is redundant and remove the above.

Reply:
will distribute subject.setUserattribute(claims) then.